### ISSUE TITLE
`let_and_return` suggestion causes error when operation between if-expression without parentheses and some value is performed in `let`-binding

### ISSUE THREAD

**Issue by zyoshoka** [rust-lang/rust-clippy#12801]
### Summary

`let_and_return` suggestion causes compilation error when operation between if-expression without parentheses and some value is performed in `let`-binding.

### Reproducer

I tried this code:

```rust
fn main() {
    get_value_str(true);
}

fn get_value_str(is_positive: bool) -> String {
    let value_str = if is_positive { "positive".to_string() } else { "negative".to_string() } + " value";
    value_str
}
```

I expected to see this happen:

```rust
fn main() {
    get_value_str(true);
}

fn get_value_str(is_positive: bool) -> String {
    (if is_positive { "positive".to_string() } else { "negative".to_string() }) + " value"
}
```

Instead, this happened:

<details><summary><code>cargo clippy --fix</code></summary>
<pre><code>    Checking rust_test v0.1.0 (/Users/username/rust-test)
warning: failed to automatically apply fixes suggested by rustc to crate `rust_test`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error: expected expression, found `+`
 --> src/main.rs:7:79
  |
7 |     if is_positive { "positive".to_string() } else { "negative".to_string() } + " value"
  |                                                                               ^ expected expression
  |
help: parentheses are required to parse this as an expression
  |
7 |     (if is_positive { "positive".to_string() } else { "negative".to_string() }) + " value"
  |     +                                                                         +

error[E0308]: mismatched types
 --> src/main.rs:7:22
  |
7 |     if is_positive { "positive".to_string() } else { "negative".to_string() } + " value"
  |     -----------------^^^^^^^^^^^^^^^^^^^^^^----------------------------------
  |     |                |
  |     |                expected `()`, found `String`
  |     expected this to be `()`
  |
help: consider using a semicolon here
  |
7 |     if is_positive { "positive".to_string(); } else { "negative".to_string() } + " value"
  |                                            +
help: consider using a semicolon here
  |
7 |     if is_positive { "positive".to_string() } else { "negative".to_string() }; + " value"
  |                                                                              +
help: you might have meant to return this value
  |
7 |     if is_positive { return "positive".to_string(); } else { "negative".to_string() } + " value"
  |                      ++++++                       +

error[E0308]: mismatched types
 --> src/main.rs:7:54
  |
7 |     if is_positive { "positive".to_string() } else { "negative".to_string() } + " value"
  |     -------------------------------------------------^^^^^^^^^^^^^^^^^^^^^^--
  |     |                                                |
  |     |                                                expected `()`, found `String`
  |     expected this to be `()`
  |
help: consider using a semicolon here
  |
7 |     if is_positive { "positive".to_string() } else { "negative".to_string(); } + " value"
  |                                                                            +
help: consider using a semicolon here
  |
7 |     if is_positive { "positive".to_string() } else { "negative".to_string() }; + " value"
  |                                                                              +
help: you might have meant to return this value
  |
7 |     if is_positive { "positive".to_string() } else { return "negative".to_string(); } + " value"
  |                                                      ++++++                       +

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0308`.
Original diagnostics will follow.

warning: returning the result of a `let` binding from a block
 --> src/main.rs:7:5
  |
6 |     let value_str = if is_positive { "positive".to_string() } else { "negative".to_string() } + " value";
  |     ----------------------------------------------------------------------------------------------------- unnecessary `let` binding
7 |     value_str
  |     ^^^^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#let_and_return
  = note: `#[warn(clippy::let_and_return)]` on by default
help: return the expression directly
  |
6 ~     
7 ~     if is_positive { "positive".to_string() } else { "negative".to_string() } + " value"
  |

warning: `rust_test` (bin "rust_test" test) generated 1 warning (run `cargo clippy --fix --bin "rust_test" --tests` to apply 1 suggestion)
warning: failed to automatically apply fixes suggested by rustc to crate `rust_test`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error: expected expression, found `+`
 --> src/main.rs:7:79
  |
7 |     if is_positive { "positive".to_string() } else { "negative".to_string() } + " value"
  |                                                                               ^ expected expression
  |
help: parentheses are required to parse this as an expression
  |
7 |     (if is_positive { "positive".to_string() } else { "negative".to_string() }) + " value"
  |     +                                                                         +

error[E0308]: mismatched types
 --> src/main.rs:7:22
  |
7 |     if is_positive { "positive".to_string() } else { "negative".to_string() } + " value"
  |     -----------------^^^^^^^^^^^^^^^^^^^^^^----------------------------------
  |     |                |
  |     |                expected `()`, found `String`
  |     expected this to be `()`
  |
help: consider using a semicolon here
  |
7 |     if is_positive { "positive".to_string(); } else { "negative".to_string() } + " value"
  |                                            +
help: consider using a semicolon here
  |
7 |     if is_positive { "positive".to_string() } else { "negative".to_string() }; + " value"
  |                                                                              +
help: you might have meant to return this value
  |
7 |     if is_positive { return "positive".to_string(); } else { "negative".to_string() } + " value"
  |                      ++++++                       +

error[E0308]: mismatched types
 --> src/main.rs:7:54
  |
7 |     if is_positive { "positive".to_string() } else { "negative".to_string() } + " value"
  |     -------------------------------------------------^^^^^^^^^^^^^^^^^^^^^^--
  |     |                                                |
  |     |                                                expected `()`, found `String`
  |     expected this to be `()`
  |
help: consider using a semicolon here
  |
7 |     if is_positive { "positive".to_string() } else { "negative".to_string(); } + " value"
  |                                                                            +
help: consider using a semicolon here
  |
7 |     if is_positive { "positive".to_string() } else { "negative".to_string() }; + " value"
  |                                                                              +
help: you might have meant to return this value
  |
7 |     if is_positive { "positive".to_string() } else { return "negative".to_string(); } + " value"
  |                                                      ++++++                       +

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0308`.
Original diagnostics will follow.

warning: `rust_test` (bin "rust_test") generated 1 warning (1 duplicate)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.18s</code></pre>
</details> 

### Version

```text
rustc 1.78.0 (9b00956e5 2024-04-29)
binary: rustc
commit-hash: 9b00956e56009bab2aa15d7bff10916599e3d6d6
commit-date: 2024-04-29
host: aarch64-apple-darwin
release: 1.78.0
LLVM version: 18.1.2
```
